pages/blog/subscriptions-and-live-queries-real-time-with-graphql.mdx

```diff
@@ -0,0 +1,1150 @@
+  })
+  .then(async (result) => {
+    if (isAsyncIterable(result)) {
+      for (const value of result) {
+        // Send to client in real-world app :)
+        console.log(value);
+      }
+    }
+  });
+
+// Invalidate by resource identifier
+rootValue.todos[0].isComplete = true;
+inMemoryLiveQueryStore.invalidate(`Todo:1`);
+
+// Invalidate by root query field coordinate
+rootValue.todos.push({ id: '2', content: 'Baz', isComplete: false });
+inMemoryLiveQueryStore.invalidate(`Query.todos`);
+```
+
+When using an ORM such as Prisma, we can simply add a middleware for automatically invalidating resources.
+
+**Use Prisma middleware for resource invalidation**
+
+```ts
+// Register Middleware for automatic model invalidation
+prisma.$use(async (params, next) => {
+  const resultPromise = next(params);
+
+  if (params.action === 'update') {
+    resultPromise.then((res) => {
+      if (res?.id) {
+        // invalidate `Post:1` on update
+        liveQueryStore.invalidate(`${params.model}:${res.id}`);
+      }
+    });
+  }
+
+  return resultPromise;
+});
+```
+
+In case we have multiple server replicas some PubSub implementation can be used for distributing the events.
+
+**PubSub with Redis**
+
+```ts
+import Redis from 'ioredis';
+
+const redis = new Redis();
+redis.subscribe('invalidations');
+redis.on('message', (channel, message) => {
+  if (channel === 'invalidations') {
+    // message is "Post:1"
+    liveQueryStore.invalidate(message);
+  }
+});
+```
+
+The transports `graphql-ws` (GraphQL over WebSocket), `graphql-helix` (GraphQL over SEE) and `@n1ru4l/socket-io-graphql-server` (GraphQL over Socket.io), support providing a custom `execute` function that is allowed to return AsyncIterables (thanks to the recent changes required for `@defer` and `@stream`).
+All we have to do is to pass the `InMemoryLiveQueryStore.execute` to our server factory!
+
+**Example with graphql-ws**
+
+```tsx
+import http from 'http';
+import ws from 'ws'; // yarn add ws
+import { useServer } from 'graphql-ws/lib/use/ws';
+import { subscribe } from 'graphql';
+import { InMemoryLiveQueryStore } from '@n1ru4l/in-memory-live-query-store';
+import { parse } from 'graphql';
+import { schema } from './schema';
+
+const inMemoryLiveQueryStore = new InMemoryLiveQueryStore();
+
+const server = http.createServer(function weServeSocketsOnly(_, res) {
+  res.writeHead(404);
+  res.end();
+});
+
+const wsServer = new ws.Server({
+  server,
+  path: '/graphql',
+});
+
+useServer(
+  {
+    schema,
+    execute: inMemoryLiveQueryStore.execute,
+    subscribe,
+  },
+  wsServer
+);
+
+server.listen(443);
+```
+
+### `@n1ru4l/graphql-live-query-patch`
+
+Optional JSON patch middleware for smaller payloads over the wire.
+
+Instead of making JSON patch required be default, it is a totally optional module, that can be applied on the client and the server for deflating (create patches on the server) and inflating (apply patches on the client) the execution results.
```

Comment:
Just a small grammatical correction: It should be `required by default`

Reply:
```suggestion
Instead of making JSON patch required by default, it is a totally optional module, that can be applied on the client and the server for deflating (create patches on the server) and inflating (apply patches on the client) the execution results.
```